src/node_main_instance.cc

```diff
@@ -0,0 +1,160 @@
+
+        if (!uv_loop_alive(env->event_loop())) {
+          EmitBeforeExit(env.get());
+        }
+
+        // Emit `beforeExit` if the loop became alive either after emitting
+        // event, or after running some callbacks.
+        more = uv_loop_alive(env->event_loop());
+      } while (more == true && !env->is_stopping());
+      env->performance_state()->Mark(
+          node::performance::NODE_PERFORMANCE_MILESTONE_LOOP_EXIT);
+    }
+
+    env->set_trace_sync_io(false);
+    exit_code = EmitExit(env.get());
+    WaitForInspectorDisconnect(env.get());
+  }
+
+  env->set_can_call_into_js(false);
+  env->stop_sub_worker_contexts();
+  uv_tty_reset_mode();
+  env->RunCleanup();
+  RunAtExit(env.get());
+
+  per_process::v8_platform.DrainVMTasks(isolate_);
+  per_process::v8_platform.CancelVMTasks(isolate_);
+
+#if defined(LEAK_SANITIZER)
+  __lsan_do_leak_check();
+#endif
+
+  return exit_code;
+}
+
+// TODO(joyeecheung): align this with the CreateEnvironment exposed in node.h
+// and the environment creation routine in workers somehow.
+std::unique_ptr<Environment> NodeMainInstance::CreateMainEnvironment(
+    int* exit_code) {
+  HandleScope handle_scope(isolate_);
+
+  // TODO(addaleax): This should load a real per-Isolate option, currently
+  // this is still effectively per-process.
+  if (isolate_data_->options()->track_heap_objects) {
+    isolate_->GetHeapProfiler()->StartTrackingHeapObjects(true);
+  }
+
+  Local<Context> context = NewContext(isolate_);
+  Context::Scope context_scope(context);
+
+  std::unique_ptr<Environment> env = std::make_unique<Environment>(
+      isolate_data_.get(),
+      context,
+      static_cast<Environment::Flags>(Environment::kIsMainThread |
+                                      Environment::kOwnsProcessState |
+                                      Environment::kOwnsInspector));
+  env->InitializeLibuv(per_process::v8_is_profiling);
+  env->ProcessCliArgs(args_, exec_args_);
+
+#if HAVE_INSPECTOR && NODE_USE_V8_PLATFORM
+  CHECK(!env->inspector_agent()->IsListening());
+  // Inspector agent can't fail to start, but if it was configured to listen
+  // right away on the websocket port and fails to bind/etc, this will return
+  // false.
+  env->inspector_agent()->Start(args_.size() > 1 ? args_[1].c_str() : "",
+                                env->options()->debug_options(),
+                                env->inspector_host_port(),
+                                true);
+  if (env->options()->debug_options().inspector_enabled &&
+      !env->inspector_agent()->IsListening()) {
+    *exit_code = 12;  // Signal internal error.
+    return env;
+  }
+#else
+  // inspector_enabled can't be true if !HAVE_INSPECTOR or
+  // !NODE_USE_V8_PLATFORM
+  // - the option parser should not allow that.
+  CHECK(!env->options()->debug_options().inspector_enabled);
+#endif  // HAVE_INSPECTOR && NODE_USE_V8_PLATFORM
+
+  if (RunBootstrapping(env.get()).IsEmpty()) {
+    *exit_code = 1;
+  }
+
+  return env;
```

Comment:
I know it's existing behavior but `*exit_code` isn't assigned on the happy path. It works because the caller sets it to zero before calling this method but still...

Reply:
It's an out parameter so at least in my understanding it's fine to not override in the happy path (think of it as overriding an default, whichever the default might be). But I guess with the current implementation resetting it to 0 at the beginning does not hurt either..